backend/src/auth/auth.service.ts

```diff
@@ -18,41 +21,79 @@ export type Tokens = {
+  email: string;
+  username: string;
+  hash: string;
+};
+
 @Injectable({})
 export class AuthService {
   constructor(
     private jwt: JwtService,
     private config: ConfigService,
-    private users: UserService
+    private users: UserService,
+    private cache: RedisCacheService,
+    private mailerService: MailerService
   ) {}
 
   /**
-   * Creates a new user object in the database and returns the JWT token
+   * Creates a new user object in the cache and send an verification email
    * @param credentials credentials validated by zod schema
-   * @returns jwt token associated with the new user
    */
-  async signup(credentials: SignupCredentialsDto): Promise<Tokens> {
+  async signup(credentials: SignupCredentialsDto) {
     const { email, username, password } = credentials;
 
-    // save new user in db
-    // const [err, user] = await this.users.find({ email, username });
-    const [err, user] = await this.users.create(email, username, password);
-
-    // if user already exists, throw exception
-    if (err) {
-      if (err instanceof PrismaClientKnownRequestError) {
-        if (err.code === "P2002") {
-          throw new ForbiddenException(AUTH_ERROR.UNAVAILABLE_CREDENTIALS);
-        }
+    // check if user's email and username in db
+    const [, user] = await this.users.findFirstByEitherUniqueFields(
+      email,
+      username
+    );
+
+    if (user) {
+      if (user.email == email) {
+        throw new ForbiddenException(AUTH_ERROR.UNAVAILABLE_EMAIL);
+      } else {
+        throw new ForbiddenException(AUTH_ERROR.UNAVAILABLE_USERNAME);
       }
-      throw err;
     }
 
-    // generate tokens for new user
-    const tokens = await this.signTokens(user.id, user.email);
-    // update refresh token hash for new user
-    await this.updateRefreshTokenHash(user.id, tokens.refresh_token);
-    return tokens;
+    const cachedEmail = await this.cache.get(email);
+    if (!!cachedEmail) {
+      throw new ForbiddenException(AUTH_ERROR.UNVERIFIED_EMAIL);
+    }
+
+    // user can be created
+    const emailVerificationToken = v4();
+    const hash = await argon2.hash(password);
+
+    // store user information in cache
+    await this.cache.set<CacheableUserFields>(emailVerificationToken, {
+      hash,
+      email,
+      username,
+    });
+    await this.cache.set<string>(email, emailVerificationToken);
+
+    // send email
+    await this.mailerService
+      .sendMail({
+        to: email,
+        subject: VERIFY_EMAIL_OPTIONS.subject,
+        template: VERIFY_EMAIL_OPTIONS.template, // The `.pug` or `.hbs` extension is appended automatically.
+        context: {
+          // Data to be sent to template engine.
+          token: emailVerificationToken,
+          username: username,
+          url: this.config.get("FRONTEND_URL"),
+        },
+      })
+      .then((success) => {
+        console.log(success);
+      })
+      .catch((err) => {
+        console.log(err);
+      });
```

Comment:
nit: can use the `radash.try` function for this? to stay consistent (plus the code easier to read)

Reply:
I had trouble using radash. Something (transporter) is undefined. I think it's okay?